repo2docker/buildpacks/pipfile/__init__.py

```diff
@@ -0,0 +1,123 @@
+        `runtime.txt`. Will return 'x.y' if version is found (e.g '3.6'), or a
+        Falsy empty string '' if not found.
+        """
+
+        if hasattr(self, '_python_version'):
+            return self._python_version
+
+        files_to_search_in_order = [
+            {
+                'path': self.binder_path('Pipfile.lock'),
+                'pattern': r'\s*\"python_(?:full_)?version\": \"?([0-9a-z\.]*)\"?', # '            "python_version": "3.6"'
+            },
+            {
+                'path': self.binder_path('Pipfile'),
+                'pattern': r'python_(?:full_)?version\s*=+\s*\"?([0-9a-z\.]*)\"?', # 'python_version = "3.6"'
+            },
+            {
+                'path': self.binder_path('runtime.txt'),
+                'pattern': r'\s*python-([0-9a-z\.]*)\s*', # 'python-3.6'
+            },
+        ]
+
+        py_version = None
+        for file in files_to_search_in_order:
+            try:
+                with open(file['path']) as f:
+                    for line in f:
+                        match = re.match(file['pattern'], line)
+                        if not match:
+                            continue
+                        py_version = match.group(1)
+                        break
+            except FileNotFoundError:
+                pass
+            if py_version:
+                break
+
+        # extract major.minor
+        if py_version:
+            if len(py_version) == 1:
+                self._python_version = self.major_pythons.get(py_version[0])
+            else:
+                # return major.minor
+                self._python_version = '.'.join(py_version.split('.')[:2])
+            return self._python_version
+        else:
+            # use the default Python
+            self._python_version = self.major_pythons['3']
+            return self._python_version
+
+    def get_assemble_scripts(self):
+        """Return series of build-steps specific to this repository.
+        """
+        # If we have a runtime.txt & that's set to python-2.7,
+        # requirements.txt will be installed in the *kernel* env
+        # and requirements3.txt (if it exists)
+        # will be installed in the python 3 notebook server env.
+        assemble_scripts = super().get_assemble_scripts()
+
+        if self.py2:
+            # using python 2 kernel,
+            # requirements3.txt allows installation in the notebook server env
+            nb_requirements_file = self.binder_path('requirements3.txt')
+            if os.path.exists(nb_requirements_file):
+                assemble_scripts.append((
+                    '${NB_USER}',
+                    # want the $NB_PYHTON_PREFIX environment variable, not for
+                    # Python's string formatting to try and replace this
+                    '${{NB_PYTHON_PREFIX}}/bin/pip install --no-cache-dir -r "{}"'.format(nb_requirements_file)
+                ))
```

Comment:
# Q2: What about this section?
Does it make sense to have like this?

Reply:
This was copied and not stripped out from the PythonBuildPack, but I'm not 100% it makes sense to have here for us, but I think it does.

Reply:
@betatim delete or keep this section?

Reply:
The `requirements3.txt` is a way to specify packages to install into the environment that your jupyter notebook server is running because if the user/kernel is Python 2 there are two environments. Having this `requirements3.txt` allows people to install server extensions which need to be in the server's env. So I'd keep that behaviour for when someone selects Python 2 in their `Pipfile`.